lib/file-source/src/paths_provider/glob.rs

```diff
@@ -0,0 +1,123 @@
+//! [`Glob`] paths provider.
+
+use super::PathsProvider;
+
+use glob::Pattern;
+use std::path::PathBuf;
+
+pub use glob::MatchOptions;
+
+/// A glob-based path provider.
+///
+/// Provides the paths to the files on the file system that match include
+/// patterns and don't match the exclude patterns.
+pub struct Glob {
+    include_patterns: Vec<String>,
+    exclude_patterns: Vec<Pattern>,
+    glob_match_options: MatchOptions,
+}
+
+impl Glob {
+    /// Create a new [`Glob`].
+    ///
+    /// Returns `None` if patterns aren't valid.
+    pub fn new(
+        include_patterns: &[PathBuf],
+        exclude_patterns: &[PathBuf],
+        glob_match_options: MatchOptions,
+    ) -> Option<Self> {
+        let include_patterns = include_patterns
+            .iter()
+            .map(|path| path.to_str().map(ToOwned::to_owned))
+            .collect::<Option<_>>()?;
+
+        let exclude_patterns = exclude_patterns
+            .iter()
+            .map(|path| path.to_str().map(|path| Pattern::new(path).ok()))
+            .flatten()
+            .collect::<Option<Vec<_>>>()?;
+
+        Some(Self {
+            include_patterns,
+            exclude_patterns,
+            glob_match_options,
+        })
+    }
+
+    /// Iterates over the paths.
+    pub fn iter(&self) -> Iter<'_> {
+        Iter {
+            include_patterns_iter: self.include_patterns.iter(),
+            exclude_patterns: self.exclude_patterns.as_slice(),
+            glob_match_options: &self.glob_match_options,
+            current_glob_iter: None,
+        }
+    }
+}
+
+/// Iterator for [`Glob`].
+pub struct Iter<'a> {
+    include_patterns_iter: std::slice::Iter<'a, String>,
+    exclude_patterns: &'a [Pattern],
+    glob_match_options: &'a glob::MatchOptions,
+    current_glob_iter: Option<glob::Paths>,
+}
+
+impl<'a> Iterator for Iter<'a> {
```

Comment:
It seems like we're taking on a decent amount of additional complexity to implement `Iterator` instead of just building a `Vec`. Given that we can only expose `Vec` via the API, why not go for a simpler implementation?

Reply:
I've pushed an alternative implementation that's more efficient than `Vec`. It has some issues though: it's definitely more complex and verbose. But the good thing is it doesn't allocate the files list in memory.
I'm considering ditching that implementation and switching back to `Vec`, but simpler, to trade-off the code simplicity for the runtime performance.
What do you think?

Reply:
While this is a very cool implementation, I don't think this functionality is performance-sensitive enough to justify the additional complexity. I would rather start with something that minimizes our maintenance overhead and then optimize if/when this code is determined to be a bottleneck.

Reply:
Sounds good! I’m going to replace the custom iter impl with a composition of combinators.